### PR TITLE
[codec] Implement asynchronous decoding source. Fixes JB#52329

### PR DESCRIPTION
**Pull request by abranson**

### CODE REVIEW

File: AsyncDecodingSource.cpp [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+    while (me->mBufferQueue.size() == 0 && !me->mReachedEOS) {
+        ALOGV("[%s] Waiting for output.", mComponentName.c_str());
+        me.waitForCondition(me->mAvailable);
+    }
+
+    if (mState == ERROR) {
+        res = ERROR;
+    }
+    else if (me->mBufferQueue.size() > 0) {
+        *buffer = *me->mBufferQueue.begin();
```
Comment (adenexter) [inline AsyncDecodingSource.cpp:229]: Ignore at your leisure, and it's probably safest that you do. But I've seen a mix of `me->mBufferQueue.size() >/= 0` and `me->mBufferQueue.empty()`, personally (as in Qt isEmpty() is the prevailing idiom) empty() is a little easier to parse.

But assuming `List` is actually a type alias of std::vector or close enough you could call `me->mBufferQueue.front()` and eliminate the iterator dereference.

File: AsyncDecodingSource.cpp [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+}
+
+status_t AsyncDecodingSource::read(
+#if ANDROID_MAJOR >= 9
+                    MediaBufferBase **buffer,
+#else
+                    MediaBuffer **buffer,
+#endif
+                    const ReadOptions *options) {
+    *buffer = NULL;
```
Comment (adenexter) [inline AsyncDecodingSource.cpp:202]: Any reason using NULL over nullptr?

**Review by adenexter — COMMENTED** [review]
To the limits of my understanding I think this looks good.
I've pointed a few idiomatic things but I realize I've probably zeroed in on existing code that's been adapted for use here.

File: AsyncDecodingSource.h [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    // returns the output format of this source.
+    virtual sp<MetaData> getFormat();
+
+    // reads from the source. This call always blocks.
+    virtual status_t read(
+#if ANDROID_MAJOR >= 9
+                    android::MediaBufferBase **buffer,
+#else
+                    android::MediaBuffer **buffer,
+#endif
```
Comment (adenexter) [inline AsyncDecodingSource.h:71]: Maybe add an type alias instead of repeating the preprocessor everywhere this turns up.

Comment (abranson) [reply AsyncDecodingSource.h]: Done for this and the MediaCodecBuffer/ABuffer too. Much cleaner.

File: AsyncDecodingSource.cpp (hunk 2 above)
Comment (abranson) [reply AsyncDecodingSource.cpp]: Just that it came from the original. Switched them all to nullptr now.

File: AsyncDecodingSource.cpp (hunk 1 above)
Comment (abranson) [reply AsyncDecodingSource.cpp:228]: There doesn't seem to be a front() in android::List, but I switched the size > 0 to !empty

File: AsyncDecodingSource.cpp [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+    if (convertMetaDataToMessage(meta, &format) != OK) {
+        return NULL;
+    }
+
+    Vector<AString> matchingCodecs;
+    MediaCodecList::findMatchingCodecs(
+            mime, false /* encoder */, flags, &matchingCodecs);
+
+    for (size_t i = 0; i < matchingCodecs.size(); ++i) {
+        const AString &componentName = matchingCodecs[i];
```
[the remark this answers is not in the document]
Comment (abranson) [reply AsyncDecodingSource.cpp]: Done

**Comment by adenexter** [comment]
LGTM

**Pull request merged**